R/data_to_long.R

```diff
@@ -0,0 +1,282 @@
+      )
+    }
+  }
+
+  if (length(names_to) > 1 && is.null(names_sep) && is.null(names_pattern)) {
+    insight::format_error(
+      "If you supply multiple names in `names_to`, you must also supply one of `names_sep` or `names_pattern`."
+    )
+  }
+
+  # Remove tidyverse attributes, will add them back at the end
+  if (inherits(data, "tbl_df")) {
+    tbl_input <- TRUE
+    data <- as.data.frame(data, stringsAsFactors = FALSE)
+  } else {
+    tbl_input <- FALSE
+  }
+
+  if (any(names_to %in% setdiff(names(data), cols))) {
+    insight::format_error(
+      "Some values of the columns specified in 'names_to' are already present as column names.",
+      paste0(
+        "Either use another value in `names_to` or rename the following columns: ",
+        text_concatenate(names_to[which(names_to %in% setdiff(names(data), cols))])
+      )
+    )
+  }
+
+  # nothing to select?
+  if (!length(cols)) {
+    stop("No columns found for reshaping data.", call. = FALSE)
+  }
+
+  not_selected <- setdiff(names(data), cols)
+
+  # create a temp id so that we know how to rearrange the rows once the data is
+  # stacked
+  not_stacked <- data[, not_selected, drop = FALSE]
+  not_stacked[["_Rows"]] <- coerce_to_numeric(row.names(data))
+
+  # stack the selected columns
+  stacked_data <- .stack(data[, cols, drop = FALSE])[, 2:1]
+
+  # reorder the rows to have a repeated sequence when all vars are selected to
+  # pivot
+  #
+  # See with following example:
+  # wide_data <- data.frame(replicate(5, rnorm(10)))
+  # data_to_long(wide_data)
+
+  needs_to_rearrange <- length(not_selected) == 0 && is.null(rows_to)
+  if (isTRUE(needs_to_rearrange)) {
+    # https://stackoverflow.com/questions/73984957/efficient-way-to-reorder-rows-to-have-a-repeated-sequence
+    stacked_data <- stacked_data[c(
+      matrix(
+        seq_len(nrow(stacked_data)),
+        nrow = length(unique(stacked_data$ind)),
+        byrow = TRUE
+      )
+    ), ]
+
+    row.names(stacked_data) <- NULL
+  }
+
+  stacked_data <- data_rename(stacked_data, "values", values_to)
+
+  # split columns if several names in names_to or names_pattern is specified
+  if (length(names_to) > 1) {
+
+    if (is.null(names_pattern)) {
+      # faster than strsplit
+      tmp <- utils::read.csv(
+        text = stacked_data$ind,
+        sep = names_sep,
+        stringsAsFactors = FALSE,
+        header = FALSE
+      )
+      names(tmp) <- paste0("V", seq_len(ncol(tmp)))
+      tmp[tmp == ""] <- NA
+
+      stacked_data$ind <- NULL
+      stacked_data <- cbind(tmp, stacked_data)
+
+    } else {
+      tmp <- regmatches(
+        unique(stacked_data$ind),
+        regexec(names_pattern, unique(stacked_data$ind))
+      )
+      tmp <- as.data.frame(do.call(rbind, tmp), stringsAsFactors = FALSE)
+      names(tmp) <- c("ind", names_to)
+      # faster than merge
+      stacked_data <- cbind(stacked_data, tmp[match(stacked_data[["ind"]], tmp[["ind"]]), -1])
```

Comment:
Maybe comment on what's the signigicance of `-1L` here? It's not obvious to me, at least.